### PR TITLE
SDIT-1698 Small refactor and add comments to difficult code

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AtttributeResolver.kt [hunk 1]
```diff
@@ -13,36 +13,15 @@ import kotlin.reflect.jvm.javaField
...
-  "aliases.middleNames",
-  "aliases.lastName",
-)
-
-class Attribute(
-  val type: KClass<*>,
-  val openSearchName: String,
-  val isNested: Boolean = false,
-  val isFuzzy: Boolean = false,
-)
```
Comment (mikehalmamoj) [inline hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AtttributeResolver.kt:38]: Moved to its own class

File: hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AtttributeResolver.kt [hunk 2]
```diff
@@ -51,37 +30,44 @@ private fun findAttributes(
...
 ): List<Pair<String, Attribute>> =
-  when (getPropertyType(prop)) {
-    PropertyType.SIMPLE -> {
-      getPropertyClass(prop).let { propClass ->
-        val propertyName = prefix + prop.name
-        val openSearchName = prefix + getOpenSearchName(prop, propClass)
-        val fuzzy = FUZZY_ATTRIBUTES.contains(propertyName)
-        listOf(propertyName to Attribute(propClass, openSearchName, nested, fuzzy))
-      }
+  when (prop.getPropertyType()) {
```
Comment (mikehalmamoj) [inline hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AtttributeResolver.kt:33]: This class is full of cryptic framework styled code, so has been littered with comments to try and give people half a chance if they ever need to change it

File: hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AtttributeResolver.kt [hunk 3]
```diff
@@ -91,22 +77,30 @@ private fun KProperty1<*, *>.getFieldType(): String? {
...
+  // A list of supported types - used only for validation to prevent people adding such lists to the Prisoner record
+  LIST_SUPPORTED_TYPES,
+
+  // A list of objects that won't have its own matcher, e.g. List<Address>
+  LIST_OBJECTS,
 }
 
-private fun getPropertyType(prop: KProperty1<*, *>): PropertyType =
-  when (prop.returnType.classifier) {
+private fun KProperty1<*, *>.getPropertyType(): PropertyType =
```
Comment (mikehalmamoj) [inline hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AtttributeResolver.kt:95]: We had a mixture of extension functions / KProperty passed as a parameter. Changed them all to extension functions for consistency.

File: hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/api/Query.kt [hunk 4]
```diff
@@ -63,13 +63,6 @@ data class Query(
       }
   }
 
-  private fun List<Matcher>.findNestedMatchers(attributes: Attributes) =
-    filterIsInstance(TypeMatcher::class.java)
-      .filter {
-        attributes[it.attribute]?.isNested
-          ?: throw AttributeSearchException("Unknown attribute: ${it.attribute}")
-      }
```
Comment (mikehalmamoj) [inline hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/api/Query.kt:71]: Moved to the Matcher class where it feels more at home.

File: hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/resource/AttributeSearchResource.kt [hunk 5]
```diff
@@ -20,18 +20,9 @@ import org.springframework.web.bind.annotation.RequestMapping
 import org.springframework.web.bind.annotation.RestController
 import uk.gov.justice.digital.hmpps.prisonersearch.search.resource.advice.ErrorResponse
 import uk.gov.justice.digital.hmpps.prisonersearch.search.services.attributesearch.AttributeSearchService
+import uk.gov.justice.digital.hmpps.prisonersearch.search.services.attributesearch.api.Attribute
 import uk.gov.justice.digital.hmpps.prisonersearch.search.services.attributesearch.api.AttributeSearchRequest
 
-@Schema(description = "An attribute that can be searched for in a query")
-data class Attribute(
```
Comment (mikehalmamoj) [inline hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/resource/AttributeSearchResource.kt:26]: Moved to its own class

File: hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AtttributeResolver.kt [hunk 6]
```diff
@@ -13,36 +13,15 @@ import kotlin.reflect.jvm.javaField
...
-
 @Configuration
 class AttributeResolver {
   @Bean
   fun attributes(): Attributes = getAttributes(Prisoner::class)
 }
 
+/**
+ * Derive all attributes that can be searched for in queries
+ */
```
Comment (mikehalmamoj) [inline hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AtttributeResolver.kt:24]: This class is full of cryptic framework style code, so has been littered with comments to try and give people half a chance if they ever need to change it

**Review by andymarke — APPROVED** [review]
lgtm

**Pull request merged**